File: crates/mdbook-html/src/html/serialize.rs
//! Serializes the [`Node`] tree to an HTML string.

use super::tree::is_void_element;
use super::tree::{Element, Node};
use ego_tree::{Tree, iter::Edge};
use html5ever::{local_name, ns};
use mdbook_core::utils::{escape_html, escape_html_attribute};
use std::ops::Deref;

/// Serializes the given tree of [`Node`] elements to an HTML string.
pub(crate) fn serialize(tree: &Tree<Node>, output: &mut String) {
    for edge in tree.root().traverse() {
        match edge {
            Edge::Open(node) => match node.value() {
                Node::Element(el) => serialize_start(el, output),
                Node::Text(text) => {
                    output.push_str(&escape_html(text));
                }
                Node::Comment(comment) => {
                    output.push_str("<!--");
                    output.push_str(comment);
                    output.push_str("-->");
                }
                Node::Fragment => {}
                Node::RawData(html) => {
                    output.push_str(html);
                }
            },
            Edge::Close(node) => {
                if let Node::Element(el) = node.value() {
                    serialize_end(el, output);
                }
            }
        }
    }
}

/// Returns true if this HTML element wants a newline to keep the emitted
/// output more readable.
fn wants_pretty_html_newline(name: &str) -> bool {
    matches!(name, |"blockquote"| "dd"
        | "div"
        | "dl"
        | "dt"
        | "h1"
        | "h2"
        | "h3"
        | "h4"
        | "h5"
        | "h6"
        | "hr"
        | "li"
        | "ol"
        | "p"
        | "pre"
        | "table"
        | "tbody"
        | "thead"
        | "tr"
        | "ul")
}

/// Emit the start tag of an element.
fn serialize_start(el: &Element, output: &mut String) {
    let el_name = el.name();
    if wants_pretty_html_newline(el_name) {
        if !output.is_empty() {
            if !output.ends_with('\n') {
                output.push('\n');
            }
        }
    }
    output.push('<');
    output.push_str(el_name);
    for (attr_name, value) in &el.attrs {
        output.push(' ');
        match attr_name.ns {
            ns!() => (),
            ns!(xml) => output.push_str("xml:"),
            ns!(xmlns) => {
                if el.name.local != local_name!("xmlns") {
                    output.push_str("xmlns:");
                }
            }
            ns!(xlink) => output.push_str("xlink:"),
            _ => (), // TODO what should it do here?
        }
        output.push_str(attr_name.local.deref());
        output.push_str("=\"");
        output.push_str(&escape_html_attribute(&value));
        output.push('"');
    }
    if el.self_closing {
        output.push_str(" /");
    }
    output.push('>');
}

/// Emit the end tag of an element.
fn serialize_end(el: &Element, output: &mut String) {
    // Void elements do not have an end tag.
    if el.self_closing || is_void_element(el.name()) {
        return;
    }
    let name = el.name();
    output.push_str("</");
    output.push_str(name);
    output.push('>');
    if wants_pretty_html_newline(name) {
        output.push('\n');
    }
}

File: mdbook-go/internal/html/serialize.go
package html

import (
	"strings"

)

// Serialize writes the tree as HTML. It is a port of
// crates/mdbook-html/src/html/serialize.rs, including its pretty-printing rule:
// a newline is emitted before the start tag of a block-level element (unless
// the output is empty or already ends in a newline) and after its end tag.
func Serialize(root *Node) string {
	var b strings.Builder
	SerializeInto(root, &b)
	return b.String()
}

// SerializeInto appends the tree to an existing buffer. The print page relies
// on this: because the pretty printer looks at what has already been written,
// serializing chapter after chapter into one buffer is what puts the page-break
// separators on their own lines.
func SerializeInto(root *Node, b *strings.Builder) {
	serializeNode(root, b)
}

func serializeNode(n *Node, b *strings.Builder) {
	switch n.Kind {
	case KindFragment:
	case KindElement:
		serializeStart(n.El, b)
	case KindText:
		b.WriteString(EscapeHTML(n.Text))
	case KindComment:
		b.WriteString("<!--")
		b.WriteString(n.Text)
		b.WriteString("-->")
	case KindRawData:
		b.WriteString(n.Text)
	}
	for _, c := range n.Children {
		serializeNode(c, b)
	}
	if n.Kind == KindElement {
		serializeEnd(n.El, b)
	}
}

// wantsPrettyHTMLNewline reports whether the element should be surrounded by
// newlines for readability.
func wantsPrettyHTMLNewline(name string) bool {
	switch name {
	case "blockquote", "dd", "div", "dl", "dt",
		"h1", "h2", "h3", "h4", "h5", "h6",
		"hr", "li", "ol", "p", "pre", "table", "tbody", "thead", "tr", "ul":
		return true
	}
	return false
}

func serializeStart(el *Element, b *strings.Builder) {
	if wantsPrettyHTMLNewline(el.Name) {
		s := b.String()
		if s != "" && !strings.HasSuffix(s, "\n") {
			b.WriteByte('\n')
		}
	}
	b.WriteByte('<')
	b.WriteString(el.Name)
	for _, attr := range el.Attrs {
		b.WriteByte(' ')
		switch attr.NS {
		case NSXML:
			b.WriteString("xml:")
		case NSXMLNS:
			if el.Name != "xmlns" {
				b.WriteString("xmlns:")
			}
		case NSXLink:
			b.WriteString("xlink:")
		}
		b.WriteString(attr.Name)
		b.WriteString("=\"")
		b.WriteString(EscapeHTMLAttribute(attr.Value))
		b.WriteByte('"')
	}
	if el.SelfClosing {
		b.WriteString(" /")
	}
	b.WriteByte('>')
}

func serializeEnd(el *Element, b *strings.Builder) {
	if el.SelfClosing || isVoidElement(el.Name) {
		return
	}
	b.WriteString("</")
	b.WriteString(el.Name)
	b.WriteByte('>')
	if wantsPrettyHTMLNewline(el.Name) {
		b.WriteByte('\n')
	}
}
